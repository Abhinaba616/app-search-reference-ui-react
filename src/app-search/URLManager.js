import createHistory from "history/createBrowserHistory";
import queryString from "query-string";
import { create } from "../types/SortOption";

function isNumeric(num) {
  return !isNaN(num);
}

function toSingleValue(val) {
  if (Array.isArray(val)) val = val[val.length - 1];
  return val;
}

function toSingleValueInteger(num) {
  return toInteger(toSingleValue(num));
}

function toInteger(num) {
  if (!isNumeric(num)) return;
  return parseInt(num, 10);
}

function parseFiltersFromQueryParams(queryParams) {
  const filters = Object.keys(queryParams).reduce((acc, paramName) => {
    if (paramName.startsWith("f-")) {
      let paramValue = queryParams[paramName];
      if (!paramValue) return acc;
      const filterName = paramName.replace("f-", "");
      acc.push({
        [filterName]: paramValue
      });
    }
    return acc;
  }, []);

  if (filters.length === 0) return;
  return filters;
}

function parseCurrentFromQueryParams(queryParams) {
  return toSingleValueInteger(queryParams.current);
}

function parseSearchTermFromQueryParams(queryParams) {
  return toSingleValue(queryParams.q);
}

function parseSortFromQueryParams(queryParams) {
  const sortBy = toSingleValue(queryParams["sort-by"]);
  const sortDirection = toSingleValue(queryParams["sort-direction"]);

  if (sortBy) {
    return create({
      value: sortBy,
      direction: sortDirection
    });
  }

  return;
}

function parseSizeFromQueryParams(queryParams) {
  return toSingleValueInteger(queryParams.size);
}

function paramsToState(queryParams) {
  const state = {
    current: parseCurrentFromQueryParams(queryParams),
    filters: parseFiltersFromQueryParams(queryParams),
    searchTerm: parseSearchTermFromQueryParams(queryParams),
    resultsPerPage: parseSizeFromQueryParams(queryParams),
    sort: parseSortFromQueryParams(queryParams)
  };

  return Object.keys(state).reduce((acc, key) => {
    const value = state[key];
    if (value) acc[key] = value;
    return acc;
  }, {});
}

function stateToParams(state) {
  const { searchTerm, current, filters, resultsPerPage, sort } = state;
  const params = {};

  filters.forEach(filter => {
    const key = Object.keys(filter)[0];
    params[`f-${key}`] = filter[key];
  });

  if (current > 1) params.current = current;
  if (searchTerm) params.q = searchTerm;
  if (resultsPerPage) params.size = resultsPerPage;
  if (sort && sort.value) {
    params["sort-by"] = sort.value;
    params["sort-direction"] = sort.direction;
  }

  return params;
}

function stateToQueryString(state) {
  const params = stateToParams(state);
  return queryString.stringify(params);
}

/**
 * The URL Manager is responsible for synchronizing state between
 * AppSearchDriver and the URL. There are 3 main cases to handle when
 * synchronizing:
 *
 * 1. When the app loads, AppSearchDriver will need to
 * read the current state from the URL, in order to perform the search
 * expressed by the query string. `getStateFromURL` is used for this case.
 *
 * 2. When the URL changes as a result of `pushState` or `replaceState`,
 * AppSearchDriver will need to be notified and given the updated state, so that
 * it can re-run the current search. `onURLStateChange` is used for this case.
 *
 * 3. When state changes in AppSearchDriver, the URL will need to be updated
 * to reflect those changes. `onURLStateChange` is used for this case.
 */

export default class URLManager {
  constructor() {
    this.history = createHistory();
  }

  getStateFromURL() {
    const state = paramsToState(
      queryString.parse(this.history.location.search)
    );
    return state;
  }

  pushStateToURL(state) {
    const searchString = stateToQueryString(state);
    this.history.push(
      {
        search: `?${searchString}`
      },
      { some: "state" }
    );
  }

  onURLStateChange() {}
}
