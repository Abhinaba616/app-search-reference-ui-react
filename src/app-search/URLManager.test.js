import URLManager from "./URLManager";

function createManager() {
  const manager = new URLManager();
  manager.history = {
    location: {
      search: ""
    },
    listen: jest.fn(),
    push: jest.fn()
  };
  return manager;
}

it("can be initialized", () => {
  const manager = new URLManager();
  expect(manager).toBeInstanceOf(URLManager);
});

describe("#getStateFromURL", () => {
  it("will parse the current state from the URL", () => {
    const manager = createManager();
    manager.history.location.search =
      "?f-dependencies=underscore&f-keywords=node&q=node&size=20&sort-direction=asc&sort-field=name";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });

  it("will ignore unrecognized parameters", () => {
    const manager = createManager();
    manager.history.location.search =
      "?q=test&bogus=12&tommy=dunn&whatever&ok&thatsall";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });

  it("will correctly handle multiple instances of the same parameter", () => {
    const manager = createManager();
    manager.history.location.search =
      "f-dependencies=underscore&f-dependencies=another&f-keywords=node&q=bad&q=node&size=bad&size=20&sort-direction=bad&sort-direction=asc&sort-field=bad&sort-field=name";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });
});

describe("#pushStateToURL", () => {
  it("will update the url with the url corresponding to the provided state", () => {
    const manager = createManager();
    manager.pushStateToURL({
      filters: [
        {
          dependencies: ["underscore", "another"]
        },
        {
          keywords: "node"
        }
      ],
      resultsPerPage: 20,
      searchTerm: "node",
      sortDirection: "asc",
      sortField: "name"
    });
    const queryString = manager.history.push.mock.calls[0][0].search;
    expect(queryString).toEqual(
      "?f-dependencies=underscore&f-dependencies=another&f-keywords=node&q=node&size=20&sort-field=name&sort-direction=asc"
    );
  });
});

describe("#onURLStateChange", () => {
  it("will call provided callback with updated state when url changes", () => {
    const manager = createManager();

    let newState;
    manager.onURLStateChange(state => {
      newState = state;
    });

    // Simulate state change
    manager.history.listen.mock.calls[0][0]({
      search:
        "?f-dependencies=underscore&f-keywords=node&q=node&size=20&sort-direction=asc&sort-field=name"
    });

    expect(newState).toMatchSnapshot();
  });
});
