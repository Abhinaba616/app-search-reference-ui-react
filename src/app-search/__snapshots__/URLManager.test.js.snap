// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getStateFromURL will correctly handle multiple instances of the same parameter 1`] = `
Object {
  "filters": Array [
    Object {
      "dependencies": Array [
        "underscore",
        "another",
      ],
    },
    Object {
      "keywords": "node",
    },
  ],
  "resultsPerPage": 20,
  "searchTerm": "node",
  "sortDirection": "asc",
  "sortField": "name",
}
`;

exports[`#getStateFromURL will ignore unrecognized parameters 1`] = `
Object {
  "searchTerm": "test",
}
`;

exports[`#getStateFromURL will parse the current state from the URL 1`] = `
Object {
  "filters": Array [
    Object {
      "dependencies": "underscore",
    },
    Object {
      "keywords": "node",
    },
  ],
  "resultsPerPage": 20,
  "searchTerm": "node",
  "sortDirection": "asc",
  "sortField": "name",
}
`;

exports[`#onURLStateChange will call provided callback with updated state when url changes 1`] = `
Object {
  "filters": Array [
    Object {
      "dependencies": "underscore",
    },
    Object {
      "keywords": "node",
    },
  ],
  "resultsPerPage": 20,
  "searchTerm": "node",
  "sortDirection": "asc",
  "sortField": "name",
}
`;
